File: old_code/LibraryService.java
package com.iriusrisk.editor.services;

import com.iriusrisk.editor.configuration.ILEConstants;
import com.iriusrisk.editor.exceptions.ILEException;
import com.iriusrisk.editor.facades.IOFacade;
import com.iriusrisk.editor.models.ILEVersion;
import com.iriusrisk.editor.models.IRBaseElement;
import com.iriusrisk.editor.models.elements.IRComponentDefinition;
import com.iriusrisk.editor.models.elements.IRLibrary;
import com.iriusrisk.editor.models.elements.IRRelation;
import com.iriusrisk.editor.models.elements.IRRiskPattern;
import com.iriusrisk.editor.models.elements.IRRiskPatternItem;
import com.iriusrisk.editor.models.elements.IRRule;
import com.iriusrisk.editor.models.elements.IRRuleAction;
import com.iriusrisk.editor.models.elements.IRRuleCondition;
import com.iriusrisk.editor.models.elements.IRThreatItem;
import com.iriusrisk.editor.models.elements.IRUseCaseItem;
import com.iriusrisk.editor.models.graph.Graph;
import com.iriusrisk.editor.models.graph.Link;
import com.iriusrisk.editor.models.graph.RuleNode;
import com.iriusrisk.editor.models.reports.IRLibraryReport;
import com.iriusrisk.editor.models.reports.IRMitigationItem;
import com.iriusrisk.editor.models.reports.IRMitigationReport;
import com.iriusrisk.editor.models.reports.IRMitigationRiskPattern;
import com.iriusrisk.editor.models.request.ComponentRequest;
import com.iriusrisk.editor.models.request.LibraryUpdateRequest;
import com.iriusrisk.editor.models.request.RelationRequest;
import com.iriusrisk.editor.models.request.RiskPatternRequest;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.springframework.stereotype.Service;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;
import java.util.function.Function;
import java.util.stream.Collectors;

import lombok.RequiredArgsConstructor;

import static java.util.stream.Collectors.toList;
import static java.util.stream.Collectors.toSet;

@Service
@RequiredArgsConstructor
public class LibraryService {

    private static final Logger logger = LogManager.getLogger(LibraryService.class);

    private final DataService dataService;
    private final IOFacade ioFacade;
    private final List<String[]> exceptions = new ArrayList<>(Arrays.asList(
            new String[]{"GENERIC-SERVICE:AUTHN-SF", "CAPEC-16"},
            new String[]{"GENERIC-SERVICE:DATA-SENS:AUTHZ", "CAPEC-232"}
    ));

    public IRLibraryReport createLibraryReport(String versionRef, String libraryRef) {
        return dataService.createLibraryReport(versionRef, libraryRef);
    }

    public void exportLibrary(String versionRef, String libraryRef, String format) {
        IRLibrary lib = dataService.getLibrary(versionRef, libraryRef);

        if ("xml".equals(format)) {

            try {
                File f = new File(ILEConstants.OUTPUT_FOLDER + "/" + versionRef);
                f.mkdirs();
                ioFacade.exportLibraryXML(lib, dataService.getVersion(versionRef), ILEConstants.OUTPUT_FOLDER + "/" + versionRef);
            } catch (Exception e) {
                throw new ILEException("Couldn't export to XML");
            }

        } else if ("xlsx".equals(format)) {
            try {
                File f = new File(ILEConstants.OUTPUT_FOLDER + "/" + versionRef);
                f.mkdirs();
                ioFacade.exportLibraryXLSX(lib, dataService.getVersion(versionRef), ILEConstants.OUTPUT_FOLDER + "/" + versionRef);
            } catch (Exception e) {
                throw new ILEException("Couldn't export to XLSX");
            }
        }

    }

    public Graph createRulesGraph(String versionRef, String libraryRef) {
        Graph g = new Graph();
        g.setDirected(true);
        g.setMultigraph(false);
        IRLibrary lib = dataService.getLibrary(versionRef, libraryRef);
        List<IRRule> rules = lib.getRules();
        TreeMap<String, IRRiskPattern> riskPatterns = lib.getRiskPatterns();
        Map<String, IRComponentDefinition> componentDefinitions = lib.getComponentDefinitions().values().stream().collect(Collectors.toMap(IRBaseElement::getRef, Function.identity()));

        for (IRRule r : rules) {
            Set<String> conditionIds = new HashSet<>();
            for (IRRuleCondition c : r.getConditions()) {
                RuleNode newCondition = new RuleNode(c.getName(), c.getValue(), "CONDITION", componentDefinitions);
                if (!g.hasNodeValue(newCondition.getId())) {
                    g.getNodes().add(newCondition);
                }

                conditionIds.add(newCondition.getId());
            }

            for (IRRuleAction a : r.getActions()) {
                RuleNode newAction = new RuleNode(a.getName(), a.getValue(), "ACTION", riskPatterns);

                if (!g.hasNodeValue(newAction.getId())) {
                    g.getNodes().add(newAction);
                }

                for (String cond : conditionIds) {
                    g.getLinks().add(new Link(cond, newAction.getId()));
                }

            }


        }


        return g;
    }

    public IRMitigationReport checkMitigation(String versionRef, String libraryRef) {
        ILEVersion version = dataService.getVersion(versionRef);
        IRLibrary lib = version.getLibrary(libraryRef);

        IRMitigationReport report = new IRMitigationReport();

        for (IRRiskPattern rp : lib.getRiskPatterns().values()) {
            IRMitigationRiskPattern s = new IRMitigationRiskPattern(rp.getRef());
            HashMap<String, IRRiskPatternItem> riskPatternTree = dataService.getRelationsInTree(lib);
            IRRiskPatternItem rpItem = riskPatternTree.get(rp.getUuid());
            
            if (rpItem != null) {
                for (IRUseCaseItem uc : rpItem.getUsecases().values()) {

                    for (IRThreatItem t : uc.getThreats().values()) {

                        // If threat is an exception we continue to the next one
                        boolean pass = false;
                        for (String[] except : exceptions) {
                            if (rp.getRef().equals(except[0]) && t.getRef().equals(except[1])) {
                                pass = true;
                                break;
                            }
                        }
                        if (pass) {
                            continue;
                        }

                        // First part: find if the threat has incorrect mitigation values
                        int mitigationCount = 0;

                        Set<IRRelation> threatRels = lib.getRelations().values().stream()
                                .filter(x -> x.getRiskPatternUuid().equals(rp.getUuid()) &&
                                        x.getUsecaseUuid().equals(uc.getRef()) &&
                                        x.getThreatUuid().equals(t.getRef()))
                                .collect(toSet());

                        Set<String> alreadyChecked = new HashSet<>();
                        for (IRRelation rel : threatRels) {
                            if (!"".equals(rel.getControlUuid()) && !alreadyChecked.contains(rel.getControlUuid())) {
                                alreadyChecked.add(rel.getControlUuid());
                                mitigationCount += Integer.parseInt(rel.getMitigation());
                            }

                        }

                        String message = "";
                        boolean error = false;

                        if (mitigationCount != 100) {
                            message = "Error with mitigation: " + mitigationCount;
                            error = true;
                        }

                        if (error) {
                            logger.info("Risk pattern: " + rp.getRef() + " -> " + t.getRef());
                            logger.info(message);
                            logger.info("Total mitigation: " + (mitigationCount));

                            IRMitigationItem item = new IRMitigationItem();
                            item.setUsecaseRef(uc.getRef());
                            item.setThreatRef(t.getRef());
                            item.setMessage(message);
                            item.setTotal("" + mitigationCount);
                            item.setRelations(threatRels);
                            item.setError(true);

                            s.getThreats().add(item);

                        }
                    }

                }
            }

            if (!s.getThreats().isEmpty()) {
                report.getRiskPatterns().add(s);
            }
        }

        return report;
    }

    /**
     * Balance mitigation values to be 100 for every threat in the library
     */
    public void balanceMitigation(String versionRef, String libraryRef) {
        ILEVersion version = dataService.getVersion(versionRef);
        IRLibrary lib = version.getLibrary(libraryRef);

        logger.info("Balancing mitigations...");
        for (IRRiskPattern rp : lib.getRiskPatterns().values()) {
            logger.info("RP:" + rp.getRef());
            HashMap<String, IRRiskPatternItem> riskPatternTree = dataService.getRelationsInTree(lib);
            IRRiskPatternItem rpItem = riskPatternTree.get(rp.getUuid());
            
            if (rpItem != null) {
                for (IRUseCaseItem uc : rpItem.getUsecases().values()) {

                    for (IRThreatItem t : uc.getThreats().values()) {

                        // If threat is an exception we continue to the next one
                        boolean pass = false;
                        for (String[] except : exceptions) {
                            if (rp.getRef().equals(except[0]) && t.getRef().equals(except[1])) {
                                pass = true;
                                break;
                            }
                        }
                        if (pass) {
                            continue;
                        }

                        // First part: find if the threat has incorrect mitigation values
                        logger.info("T: " + t.getRef());
                        List<IRRelation> relationList = new ArrayList<>();
                        List<IRRelation> threatRels = lib.getRelations().values().stream()
                                .filter(x -> x.getRiskPatternUuid().equals(rp.getUuid()) &&
                                        x.getUsecaseUuid().equals(uc.getRef()) &&
                                        x.getThreatUuid().equals(t.getRef())).collect(toList());

                        for (IRRelation rel : threatRels) {
                            if (!"".equals(rel.getControlUuid())) {
                                relationList.add(rel);
                            }
                        }

                        fixMitigationValues(relationList, 100);
                    }

                }
            }
        }

        logger.info("Balanced!");
    }

    private void fixMitigationValues(List<IRRelation> all, int goal) {
        if (!all.isEmpty()) {
            // Check first if the mitigation sum the given value
            int mitigationSum = 0;
            for (IRRelation rel : all) {
                mitigationSum += Integer.parseInt(rel.getMitigation());
            }

            if (mitigationSum != goal) {
                logger.info("Sum is " + mitigationSum + " (Desired: " + goal + "). Fixing threat...");
                int mean = goal / all.size();
                int remainder = goal % all.size();
                for (IRRelation rel : all) {
                    int newMit = mean;

                    if (remainder != 0) {
                        newMit += remainder;
                        remainder = 0;
                    }

                    if (!rel.getMitigation().equals("" + newMit)) {
                        logger.info("Control: Updated mitigation for " + rel.getControlUuid() + ": " + rel.getMitigation() + " -> " + newMit);
                        rel.setMitigation("" + newMit);
                    } else {
                        logger.info("No changes for " + rel.getControlUuid());
                    }
                }
            }
        }

    }




    // Libraries

    public void updateLibrary(String versionRef, String libraryRef, LibraryUpdateRequest newLib) {
        IRLibrary currentLib = dataService.getLibrary(versionRef, libraryRef);
        currentLib.setName(newLib.getName());
        currentLib.setDesc(newLib.getDesc());
        currentLib.setRevision(newLib.getRevision());
        currentLib.setFilename(newLib.getFilename());
        currentLib.setEnabled(newLib.getEnabled());
    }

    // Components

    public Collection<IRComponentDefinition> listComponents(String versionRef, String library) {
        return dataService.getLibrary(versionRef, library).getComponentDefinitions().values();
    }

    public IRComponentDefinition addComponent(String versionRef, String lib, ComponentRequest body) {
        ILEVersion v = dataService.getVersion(versionRef);
        IRLibrary l = v.getLibrary(lib);
        IRComponentDefinition comp = new IRComponentDefinition(body.getRef(), body.getName(), body.getDesc(), body.getCategoryRef(), body.getVisible());
        l.getComponentDefinitions().put(comp.getUuid(), comp);
        return comp;
    }

    public IRComponentDefinition updateComponent(String versionRef, String lib, IRComponentDefinition newComp) {
        ILEVersion v = dataService.getVersion(versionRef);
        IRLibrary l = v.getLibrary(lib);
        l.getComponentDefinitions().put(newComp.getUuid(), newComp);
        return newComp;
    }

    public void deleteComponent(String versionRef, String lib, IRComponentDefinition comp) {
        ILEVersion v = dataService.getVersion(versionRef);
        IRLibrary l = v.getLibrary(lib);
        l.getComponentDefinitions().remove(comp.getUuid());
    }

    // Risk patterns

    public Collection<IRRiskPattern> listRiskPatterns(String versionRef, String library) {
        return dataService.getLibrary(versionRef, library).getRiskPatterns().values();
    }

    public IRRiskPattern addRiskPattern(String versionRef, String libraryRef, RiskPatternRequest request) {
        ILEVersion v = dataService.getVersion(versionRef);
        IRLibrary l = v.getLibrary(libraryRef);
        IRRiskPattern rp = new IRRiskPattern(request.getRef(), request.getName(), request.getDesc());
        l.getRiskPatterns().put(rp.getUuid(), rp);
        return rp;
    }

    public IRRiskPattern updateRiskPattern(String versionRef, String lib, RiskPatternRequest newRp) {
        ILEVersion v = dataService.getVersion(versionRef);
        IRLibrary l = v.getLibrary(lib);
        IRRiskPattern rp = l.getRiskPatterns().get(newRp.getUuid());
        if(newRp.getRef()!=null){
            rp.setRef(newRp.getRef());
        }

        if(newRp.getName()!=null){
            rp.setName(newRp.getName());
        }

        if(newRp.getDesc()!=null){
            rp.setDesc(newRp.getDesc());
        }

        l.getRiskPatterns().put(rp.getUuid(), rp);
        return rp;
    }

    public void deleteRiskPattern(String versionRef, String lib, IRRiskPattern rp) {
        ILEVersion v = dataService.getVersion(versionRef);
        IRLibrary l = v.getLibrary(lib);
        l.getRiskPatterns().remove(rp.getUuid());
    }

    // Relations

    public Collection<IRRelation> listRelations(String versionRef, String library) {
        return dataService.getLibrary(versionRef, library).getRelations().values();
    }

    public IRRelation addRelation(String versionRef, String lib, RelationRequest body) {
        ILEVersion v = dataService.getVersion(versionRef);
        IRLibrary l = v.getLibrary(lib);
        IRRelation rel = new IRRelation(body.getRiskPatternUuid(), body.getUsecaseUuid(), body.getThreatUuid(), body.getWeaknessUuid(), body.getControlUuid(), body.getMitigation());
        l.getRelations().put(rel.getUuid(), rel);
        return rel;
    }

    public IRRelation updateRelation(String versionRef, String lib, IRRelation newRel) {
        ILEVersion v = dataService.getVersion(versionRef);
        IRLibrary l = v.getLibrary(lib);
        l.getRelations().put(newRel.getUuid(), newRel);
        return newRel;
    }

    public void deleteRelation(String versionRef, String lib, IRRelation rel) {
        ILEVersion v = dataService.getVersion(versionRef);
        IRLibrary l = v.getLibrary(lib);
        l.getRelations().remove(rel.getUuid());
    }

}

File: isra/src/ile/backend/app/services/library_service.py
"""
Library service for IriusRisk Library Editor API
"""

import logging
import os
from pathlib import Path
from typing import Collection, Dict, List, Set

from isra.src.ile.backend.app.configuration.constants import ILEConstants
from isra.src.ile.backend.app.models import (
    ILEVersion, IRBaseElement, IRComponentDefinition, IRLibrary,
    IRRelation, IRRiskPattern, IRRiskPatternItem, IRRule,
    IRRuleAction, IRRuleCondition, IRThreatItem, IRUseCaseItem,
    Graph, Link, RuleNode, IRLibraryReport, IRMitigationItem,
    IRMitigationReport, IRMitigationRiskPattern, ComponentRequest,
    LibraryUpdateRequest, RelationRequest, RiskPatternRequest
)
from isra.src.ile.backend.app.services.data_service import DataService
from isra.src.ile.backend.app.facades.io_facade import IOFacade

logger = logging.getLogger(__name__)


class LibraryService:
    """Service for handling library operations"""
    
    def __init__(self):
        self.data_service = DataService()
        self.io_facade = IOFacade()
        self.exceptions = [
            ["GENERIC-SERVICE:AUTHN-SF", "CAPEC-16"],
            ["GENERIC-SERVICE:DATA-SENS:AUTHZ", "CAPEC-232"]
        ]
    
    def create_library_report(self, version_ref: str, library_ref: str) -> IRLibraryReport:
        """Create library report"""
        return self.data_service.create_library_report(version_ref, library_ref)
    
    def export_library(self, version_ref: str, library_ref: str, format: str) -> None:
        """Export library to specified format"""
        lib = self.data_service.get_library(version_ref, library_ref)
        
        if format == "xml":
            try:
                output_path = Path(ILEConstants.OUTPUT_FOLDER) / version_ref
                output_path.mkdir(parents=True, exist_ok=True)
                self.io_facade.export_library_xml(lib, self.data_service.get_version(version_ref), str(output_path))
            except Exception as e:
                raise RuntimeError("Couldn't export to XML") from e
        elif format == "xlsx":
            try:
                output_path = Path(ILEConstants.OUTPUT_FOLDER) / version_ref
                output_path.mkdir(parents=True, exist_ok=True)
                self.io_facade.export_library_xlsx(lib, self.data_service.get_version(version_ref), str(output_path))
            except Exception as e:
                raise RuntimeError("Couldn't export to XLSX") from e
    
    def create_rules_graph(self, version_ref: str, library_ref: str) -> Graph:
        """Create rules graph"""
        g = Graph()
        g.directed = True
        g.multigraph = False
        
        lib = self.data_service.get_library(version_ref, library_ref)
        rules = lib.rules
        risk_patterns = lib.risk_patterns
        component_definitions = {comp.ref: comp for comp in lib.component_definitions.values()}
        
        for r in rules:
            condition_ids = set()
            for c in r.conditions:
                new_condition = RuleNode(c.name, c.value, "CONDITION", component_definitions)
                if not g.has_node_value(new_condition.id):
                    g.nodes.append(new_condition)
                
                condition_ids.add(new_condition.id)
            
            for a in r.actions:
                new_action = RuleNode(a.name, a.value, "ACTION", risk_patterns)
                
                if not g.has_node_value(new_action.id):
                    g.nodes.append(new_action)
                
                for cond in condition_ids:
                    g.links.append(Link(cond, new_action.id))
        
        return g
    
    def check_mitigation(self, version_ref: str, library_ref: str) -> IRMitigationReport:
        """Check mitigation values"""
        version = self.data_service.get_version(version_ref)
        lib = version.get_library(library_ref)
        
        report = IRMitigationReport()
        
        for rp in lib.risk_patterns.values():
            s = IRMitigationRiskPattern(risk_pattern_ref=rp.ref)
            risk_pattern_tree = self.data_service.get_relations_in_tree(lib)
            rp_item = risk_pattern_tree.get(rp.uuid)
            
            if rp_item is not None:
                for uc in rp_item.usecases.values():
                    for t in uc.threats.values():
                        # If threat is an exception we continue to the next one
                        pass_exception = False
                        for except_item in self.exceptions:
                            if rp.ref == except_item[0] and t.ref == except_item[1]:
                                pass_exception = True
                                break
                        if pass_exception:
                            continue
                        
                        # First part: find if the threat has incorrect mitigation values
                        mitigation_count = 0
                        
                        threat_rels = [
                            rel for rel in lib.relations.values()
                            if (rel.risk_pattern_uuid == rp.uuid and
                                rel.usecase_uuid == uc.ref and
                                rel.threat_uuid == t.ref)
                        ]
                        
                        already_checked = set()
                        for rel in threat_rels:
                            if rel.control_uuid != "" and rel.control_uuid not in already_checked:
                                already_checked.add(rel.control_uuid)
                                mitigation_count += int(rel.mitigation)
                        
                        message = ""
                        error = False
                        
                        if mitigation_count != 100:
                            message = f"Error with mitigation: {mitigation_count}"
                            error = True
                        
                        if error:
                            logger.info(f"Risk pattern: {rp.ref} -> {t.ref}")
                            logger.info(message)
                            logger.info(f"Total mitigation: {mitigation_count}")
                            
                            item = IRMitigationItem(
                                usecase_ref=uc.ref,
                                threat_ref=t.ref,
                                message=message,
                                total=str(mitigation_count),
                                relations=threat_rels,
                                error=True
                            )
                            
                            s.threats.append(item)
            
            if s.threats:
                report.risk_patterns.append(s)
        
        return report
    
    def balance_mitigation(self, version_ref: str, library_ref: str) -> None:
        """Balance mitigation values to be 100 for every threat in the library"""
        version = self.data_service.get_version(version_ref)
        lib = version.get_library(library_ref)
        
        logger.info("Balancing mitigations...")
        for rp in lib.risk_patterns.values():
            logger.info(f"RP: {rp.ref}")
            risk_pattern_tree = self.data_service.get_relations_in_tree(lib)
            rp_item = risk_pattern_tree.get(rp.uuid)
            
            if rp_item is not None:
                for uc in rp_item.usecases.values():
                    for t in uc.threats.values():
                        # If threat is an exception we continue to the next one
                        pass_exception = False
                        for except_item in self.exceptions:
                            if rp.ref == except_item[0] and t.ref == except_item[1]:
                                pass_exception = True
                                break
                        if pass_exception:
                            continue
                        
                        logger.info(f"T: {t.ref}")
                        relation_list = []
                        threat_rels = [
                            rel for rel in lib.relations.values()
                            if (rel.risk_pattern_uuid == rp.uuid and
                                rel.usecase_uuid == uc.ref and
                                rel.threat_uuid == t.ref)
                        ]
                        
                        for rel in threat_rels:
                            if rel.control_uuid != "":
                                relation_list.append(rel)
                        
                        self._fix_mitigation_values(relation_list, 100)
        
        logger.info("Balanced!")
    
    def _fix_mitigation_values(self, all_relations: List[IRRelation], goal: int) -> None:
        """Fix mitigation values to reach goal"""
        if not all_relations:
            return
        
        # Check first if the mitigation sum the given value
        mitigation_sum = sum(int(rel.mitigation) for rel in all_relations)
        
        if mitigation_sum != goal:
            logger.info(f"Sum is {mitigation_sum} (Desired: {goal}). Fixing threat...")
            mean = goal // len(all_relations)
            remainder = goal % len(all_relations)
            
            for rel in all_relations:
                new_mit = mean
                
                if remainder != 0:
                    new_mit += remainder
                    remainder = 0
                
                if rel.mitigation != str(new_mit):
                    logger.info(f"Control: Updated mitigation for {rel.control_uuid}: {rel.mitigation} -> {new_mit}")
                    rel.mitigation = str(new_mit)
                else:
                    logger.info(f"No changes for {rel.control_uuid}")
    
    def update_library(self, version_ref: str, library_ref: str, new_lib: LibraryUpdateRequest) -> None:
        """Update library"""
        current_lib = self.data_service.get_library(version_ref, library_ref)
        current_lib.name = new_lib.name
        current_lib.desc = new_lib.desc
        current_lib.revision = new_lib.revision
        current_lib.filename = new_lib.filename
        current_lib.enabled = new_lib.enabled
    
    def list_components(self, version_ref: str, library: str) -> Collection[IRComponentDefinition]:
        """List components"""
        return self.data_service.get_library(version_ref, library).component_definitions.values()
    
    def add_component(self, version_ref: str, lib: str, body: ComponentRequest) -> IRComponentDefinition:
        """Add component"""
        v = self.data_service.get_version(version_ref)
        l = v.get_library(lib)
        comp = IRComponentDefinition(
            ref=body.ref,
            name=body.name,
            desc=body.desc,
            category_ref=body.category_ref,
            visible=body.visible
        )
        l.component_definitions[comp.uuid] = comp
        return comp
    
    def update_component(self, version_ref: str, lib: str, new_comp: IRComponentDefinition) -> IRComponentDefinition:
        """Update component"""
        v = self.data_service.get_version(version_ref)
        l = v.get_library(lib)
        l.component_definitions[new_comp.uuid] = new_comp
        return new_comp
    
    def delete_component(self, version_ref: str, lib: str, comp: IRComponentDefinition) -> None:
        """Delete component"""
        v = self.data_service.get_version(version_ref)
        l = v.get_library(lib)
        l.component_definitions.pop(comp.uuid, None)
    
    def list_risk_patterns(self, version_ref: str, library: str) -> Collection[IRRiskPattern]:
        """List risk patterns"""
        return self.data_service.get_library(version_ref, library).risk_patterns.values()
    
    def add_risk_pattern(self, version_ref: str, library_ref: str, request: RiskPatternRequest) -> IRRiskPattern:
        """Add risk pattern"""
        v = self.data_service.get_version(version_ref)
        l = v.get_library(library_ref)
        rp = IRRiskPattern(
            ref=request.ref,
            name=request.name,
            desc=request.desc
        )
        l.risk_patterns[rp.uuid] = rp
        return rp
    
    def update_risk_pattern(self, version_ref: str, lib: str, new_rp: RiskPatternRequest) -> IRRiskPattern:
        """Update risk pattern"""
        v = self.data_service.get_version(version_ref)
        l = v.get_library(lib)
        rp = l.risk_patterns[new_rp.uuid]
        
        if new_rp.ref is not None:
            rp.ref = new_rp.ref
        
        if new_rp.name is not None:
            rp.name = new_rp.name
        
        if new_rp.desc is not None:
            rp.desc = new_rp.desc
        
        l.risk_patterns[rp.uuid] = rp
        return rp
    
    def delete_risk_pattern(self, version_ref: str, lib: str, rp: IRRiskPattern) -> None:
        """Delete risk pattern"""
        v = self.data_service.get_version(version_ref)
        l = v.get_library(lib)
        l.risk_patterns.pop(rp.uuid, None)
    
    def list_relations(self, version_ref: str, library: str) -> Collection[IRRelation]:
        """List relations"""
        return self.data_service.get_library(version_ref, library).relations.values()
    
    def add_relation(self, version_ref: str, lib: str, body: RelationRequest) -> IRRelation:
        """Add relation"""
        v = self.data_service.get_version(version_ref)
        l = v.get_library(lib)
        rel = IRRelation(
            risk_pattern_uuid=body.risk_pattern_uuid,
            usecase_uuid=body.usecase_uuid,
            threat_uuid=body.threat_uuid,
            weakness_uuid=body.weakness_uuid,
            control_uuid=body.control_uuid,
            mitigation=body.mitigation
        )
        l.relations[rel.uuid] = rel
        return rel
    
    def update_relation(self, version_ref: str, lib: str, new_rel: IRRelation) -> IRRelation:
        """Update relation"""
        v = self.data_service.get_version(version_ref)
        l = v.get_library(lib)
        l.relations[new_rel.uuid] = new_rel
        return new_rel
    
    def delete_relation(self, version_ref: str, lib: str, rel: IRRelation) -> None:
        """Delete relation"""
        v = self.data_service.get_version(version_ref)
        l = v.get_library(lib)
        l.relations.pop(rel.uuid, None)
